hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/processors/StreamToStreamJoinP.java

```diff
@@ -0,0 +1,262 @@
+        this.leftTimeExtractors = leftTimeExtractors;
+        this.rightTimeExtractors = rightTimeExtractors;
+        this.postponeTimeMap = postponeTimeMap;
+        this.columnCount = columnCount;
+
+        for (byte key : postponeTimeMap.keySet()) {
+            Map<Byte, Long> valueMap = postponeTimeMap.get(key);
+            Map<Byte, Long> map = new HashMap<>();
+            for (byte j : valueMap.keySet()) {
+                map.put(j, Long.MIN_VALUE);
+            }
+            wmState.put(key, map);
+        }
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) throws Exception {
+        this.evalContext = ExpressionEvalContext.from(context);
+    }
+
+    @SuppressWarnings("ConstantConditions")
+    @Override
+    public boolean tryProcess(int ordinal, @Nonnull Object item) {
+        assert ordinal == 0 || ordinal == 1; // bad DAG
+
+        while (!pendingOutput.isEmpty()) {
+            if (!tryEmit(pendingOutput.peek())) {
+                return false;
+            } else {
+                pendingOutput.remove();
+            }
+        }
+
+        //  having side input, traverse opposite buffer
+        if (currItem == null) {
+            currItem = (JetSqlRow) item;
+            buffer[ordinal].add(currItem);
+            iterator = buffer[1 - ordinal].iterator();
+            if (!joinInfo.isInner()) {
+                unusedEventsTracker[ordinal].add(currItem);
+            }
+        }
+
+        if (!iterator.hasNext()) {
+            iterator = null;
+            currItem = null;
+            return true;
+        }
+
+        JetSqlRow oppositeBufferItem = iterator.next();
+        JetSqlRow preparedOutput = ExpressionUtil.join(
+                ordinal == 0 ? currItem : oppositeBufferItem,
+                ordinal == 0 ? oppositeBufferItem : currItem,
+                joinInfo.condition(),
+                evalContext
+        );
+        // it is used already once
+        unusedEventsTracker[1 - ordinal].remove(oppositeBufferItem);
+
+        if (preparedOutput != null && !tryEmit(preparedOutput)) {
+            pendingOutput.offer(preparedOutput);
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean tryProcessWatermark(int ordinal, @Nonnull Watermark watermark) {
+        // if pending watermarks available - try to send them
+        if (pendingWatermark != null && !tryEmit(pendingWatermark)) {
+            return false;
+        }
+        pendingWatermark = null;
+
+        // update wm state
+        applyToWmState(watermark);
+
+        // try to clear buffers if possible
+        clearExpiredItemsInBuffer(ordinal);
+
+        // We can't immediately emit current WM, as it could render items in buffers late.
+        // Instead, we can emit WM with the minimum available time for this WM key.
+        byte wmKey = watermark.key();
+        long minItemTime = findMinimumBufferTime(ordinal, wmKey);
+        if (minItemTime != MAX_VALUE) {
```

Comment:
Multiple issues here:
1. this algorithm can emit the same WM twice. We must not compare against `Long.MIN_VALUE`, but against the last emitted value
2. watermark received for key0 could trigger a new wm for key1 to have a new minimum through changes in `wmState`, we don't handle it
3. if there's nothing in the buffer (`findMinimumBufferTime` returns `MAX_VALUE`), we must emit the WM according to the minimum group time